pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go

```diff
@@ -210,19 +215,99 @@ func (state PipelineRunState) GetNextTasks(candidateTasks map[string]struct{}) [
-	done := []string{}
+// SuccessfulOrSkippedDAGTasks returns a list of the names of all of the PipelineTasks in state
+// which have successfully completed or skipped
+func (state PipelineRunState) SuccessfulOrSkippedDAGTasks(d *dag.Graph) []string {
+	tasks := []string{}
+	stateMap := state.ToMap()
 	for _, t := range state {
-		if t.TaskRun != nil {
-			c := t.TaskRun.Status.GetCondition(apis.ConditionSucceeded)
-			if c.IsTrue() {
-				done = append(done, t.PipelineTask.Name)
+		if _, ok := d.Nodes[t.PipelineTask.Name]; ok {
+			if t.IsSuccessful() || isSkipped(t, stateMap, d) {
+				tasks = append(tasks, t.PipelineTask.Name)
+			}
+		}
+	}
+	return tasks
+}
+
+// isDAGTasksStopped returns true if any of the pipeline task has failed
+// and none of the pipeline task are still running
+func (state PipelineRunState) isDAGTasksStopped(d *dag.Graph) bool {
+	failed := false
+	for _, t := range state {
+		if t.IsFailure() {
+			failed = true
+			continue
+		}
+		if t.IsStarted() && !t.IsDone() {
+			failed = false
+			break
+		}
+	}
+	return failed
+}
+
+// isDAGTasksDone returns true if all DAG tasks are done executing,
+// DAG tasks could have failed/succeeded/skipped
+func (state PipelineRunState) isDAGTasksDone(d *dag.Graph) (isDone bool) {
+	isDone = true
+	stateMap := state.ToMap()
+	for _, t := range state {
+		if _, ok := d.Nodes[t.PipelineTask.Name]; ok {
+			if t.TaskRun == nil {
+				// this task might have skipped if taskRun is nil
+				// continue and ignore if this task was skipped
+				// skipped task is considered part of done
+				if isSkipped(t, stateMap, d) {
+					continue
+				}
+				return false
+			}
+			isDone = isDone && t.IsDone()
+			if !isDone {
+				return
+			}
+		}
+	}
+	return
+}
+
+// isFinalTasksDone returns true if all final tasks are done executing,
+// final tasks could have failed/succeeded
+func (state PipelineRunState) isFinalTasksDone(d *dag.Graph) (isDone bool) {
+	isDone = true
+	for _, t := range state {
+		if _, ok := d.Nodes[t.PipelineTask.Name]; ok {
+			if t.TaskRun == nil {
+				return false
+			}
+			isDone = isDone && t.IsDone()
+			if !isDone {
+				return
 			}
 		}
 	}
-	return done
+	return
+}
+
+// GetFinalTasks returns a list of final tasks without any taskRun associated with it
+// GetFinalTasks returns final tasks only when all DAG tasks have finished executing successfully or skipped or
+// any one DAG task resulted in failure
+func (state PipelineRunState) GetFinalTasks(d *dag.Graph, dfinally *dag.Graph) []*ResolvedPipelineRunTask {
+	tasks := []*ResolvedPipelineRunTask{}
+	finalCandidates := map[string]struct{}{}
+	// check either pipeline has finished executing all DAG pipelineTasks
+	// or any one of the DAG pipelineTask has failed
+	if state.isDAGTasksStopped(d) || state.isDAGTasksDone(d) {
+		// return list of tasks with all final tasks
+		for _, t := range state {
+			if _, ok := dfinally.Nodes[t.PipelineTask.Name]; ok && !t.IsSuccessful() {
```

Comment:
i dont quite understand why we are checking if the tasks is not successful - i would have thought we want to check if the task has started?

Reply:
these checks are done to build potential candidates for execution, `!t.IsSuccessful()` is making a final task is not excluded until it's successful. The task could have started but might fail, if it fails, we want to rerun it until all retry attempts are exhausted similar to DAG task.